### DISCUSSION TITLE
Activity Feeds from custom schemas

### DISCUSSION THREAD

**Discussion by alexborton** [bluesky-social/atproto#2827] [How to do it?]
This is a broad "how to" so i can fill some knowledge gaps and work out the feasibility for a green field project - please excuse any naivety and dumb questions!

I am in the process of (re)building a social media application. In short, I would like to float all the social interactions into the Atproto. For simplicity, imagine a Rating/review platform whereby users have Friends (follows) that they trust for reviews.

Authentication and custom schemas to set this data for individual users seems straight forward enough (the "statusphere" example does most of this!). What i don't know from a feasibility and technical perspective is what the "activity feed" (or "feed generator") looks like and how that might work.

Specifically, can i create a feed for individual users that surfaces Rating/review (custom schema) events from their follows only, on specific schemas? And what does that look like from a basic, technical perspective. 

🙏

**Comment by bnewbold** [comment]
This sounds like a great project! I think reviews with a social aspect are a really good fit for atproto.

My understanding of your question is that you are looking for advice on how to implement "timelines" or "feeds" in your app, where the content of the "recent reviews" list has different content for each viewing user, because each user has different social connections (eg, "follows", "friends") which influence what shows up. Is that right?

This is a classic design/architecture challenge. It comes up a lot in atproto projects, but is also relevant and discussed for central projects. The classic reverse-chronological Twitter timeline, which depends only on your follow graph, is the classic example of the pattern.

The short answer is that this is a "view" of content in the network, and you'll need to implement it as part of your "AppView" (API service), eg extend the statusphere example to have database tables, indexing, and queries to implement this. 

There are two broad approaches:

- query-driven timeline assembly, where database queries when the view is requested to find relevant items (eg, reviews) from relevant accounts (based on relationship) for the given viewer account. These could be SQL JOINs, multiple queries, nested sub-queries, etc. this is simple and works well up to fairly large scale (eg, a million users or more); it gets harder to serve the read queries when the database gets very large.
- "fan out" of feed content at content creation time. basically keep track of timelines for each user in the network: when a new item (eg, review) is created, append it to all the relevant timelines of accounts which follow the author. handling all the social graph changes (adding/removing follows, etc) is complex. this is common at very large scale, because it makes "get timeline" very fast, but requires a lot of resources.

I'd recommend starting with the first (query-driven lookups) until the size of your database is very large (eg, hundreds of millions of reviews).
